File: Assignment 3/BinaryTree/Java/BinaryTree.java
/**
 * A runnable class that implements and tests an ordered binary tree of integer values.
 */
public class BinaryTree {

  /**
   * A node of the binary tree containing the node's integer value 
   * and pointers to its right and left children (or null).
   */
  static class Node {
    int  value;
    Node left   = null;
    Node right  = null;
    
    /**
     * Create a new node with no children.
     */
    Node (int value) {
      this.value = value;
    }
    
    /**
     * Insert the node n into the binary tree rooted by this node.
     */
    void insertNode (Node n) {
      if (n.value <= value) {
        if (left==null)
          left = n;
        else
          left.insertNode (n);
      } else {
        if (right==null)
          right = n;
        else
          right.insertNode (n);
      }
    }
    
    /**
     * Insert value into the binary tree rooted by this node, creating a new node for value.
     */
    void insert (int value) {
      Node n = new Node (value);
      insertNode (n);
    }
    
    /**
     * Print the contents entire binary tree in order of ascending integer value.
     */
    void printInOrder() {
      if (left != null)
        left.printInOrder();
      System.out.printf ("%d\n", value);
      if (right != null)
        right.printInOrder();
    }
    
  }
  
  /**
   * Create a node to root a new tree, populate it with a few values, and print it.
   */
  public static void main (String[] args) {
    Node root = new Node (100);
    root.insert (10);
    root.insert (120);
    root.insert (130);
    root.insert (90);
    root.insert (5);
    root.insert (95);
    root.insert (121);
    root.insert (131);
    root.insert (1);
    root.printInOrder();
  }

}

File: Assignment 3/BinaryTree/C/BinaryTree.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>

//
// A binary tree node with value and left and right children
struct Node {
  int value;
  struct Node *left;
  struct Node *right;
};

//
// Insert node n into tree rooted at toNode
//
void insertNode (struct Node* toNode, struct Node* n) {
  if (n->value <= toNode->value) {
    if (toNode->left == NULL) {
      toNode->left = n;
    }
    else {
      insertNode(toNode->left, n);
    }
  }
  else {
    if (toNode->right == NULL) {
      toNode->right = n;
    }
    else {
      insertNode(toNode->right, n);
    }
  }
}

//
// Insert new node with specified value into tree rooted at toNode
//
void insert (struct Node* toNode, int value) {
  struct Node *newNode = malloc(sizeof(struct Node));
  newNode->value = value;
  insertNode(toNode, newNode);
}


//
// Print values of tree rooted at node in ascending order
//
void printInOrder (struct Node* node) {
  if (node->left != NULL) {
    printInOrder(node->left);
  }
  printf("%d\n", node->value);
  if (node->right != NULL) {
    printInOrder(node->right);
  }
}

//
// Create root node, insert some values, and print tree in order
//
int main (void) {
  struct Node root = {100, NULL, NULL};
  insert(&root, 10);
  insert(&root, 120);
  insert(&root, 130);
  insert(&root, 90);
  insert(&root, 5);
  insert(&root, 95);
  insert(&root, 121);
  insert(&root, 131);
  insert(&root, 1);
  printInOrder(&root);
  return 0;
}
